### ISSUE TITLE
Producer: panic - concurrent map read and map write

### ISSUE THREAD

**Issue by bzz** [src-d/borges#178]
Happened while fetching 30k repos

```
t=2017-10-31T15:46:41+0000 lvl=dbug msg="Copy siva file to HDFS" module=borges worker=0 job=015f72dd-c378-f359-8d34-0f77b4d075e6 endpoint=git://github.com/libpd/puredatap5.git root=3e0dbce3c723e7c65869d33f264873c62cb9d937 RootedRepository=3e0dbce3c723e7c65869d33f264873c62cb9d937 copyToRemote=0 caller=archiver.go:311
fatal error: concurrent map read and map write

goroutine 49 [running]:
runtime.throw(0xcd6561, 0x21)
	/home/travis/.gimme/versions/go/src/runtime/panic.go:606 +0x81 fp=0xc9778b30f8 sp=0xc9778b30d8 pc=0x42b341
runtime.mapaccess2_faststr(0xbb24c0, 0xc42012ca80, 0xc52f629620, 0xf, 0x45, 0x3a)
	/home/travis/.gimme/versions/go/src/runtime/hashmap_fast.go:270 +0x46e fp=0xc9778b3168 sp=0xc9778b30f8 pc=0x40b98e
github.com/src-d/borges/vendor/gopkg.in/src-d/go-git.v4/plumbing/transport/client.NewClient(0xd69840, 0xc4202f3080, 0xd69840, 0xc4202f3080, 0x0, 0x0)
	/home/travis/gopath/src/github.com/src-d/borges/vendor/gopkg.in/src-d/go-git.v4/plumbing/transport/client/client.go:38 +0x6e fp=0xc9778b31f0 sp=0xc9778b3168 pc=0x8c0cde
github.com/src-d/borges/vendor/gopkg.in/src-d/go-git%2ev4.newClient(0xc52f629620, 0x16, 0x0, 0x0, 0x12, 0x13, 0x0, 0xc9778b32a0)
	/home/travis/gopath/src/github.com/src-d/borges/vendor/gopkg.in/src-d/go-git.v4/remote.go:333 +0x6e fp=0xc9778b3240 sp=0xc9778b31f0 pc=0x8cadfe
github.com/src-d/borges/vendor/gopkg.in/src-d/go-git%2ev4.newSendPackSession(0xc52f629620, 0x16, 0x0, 0x0, 0x0, 0x0, 0xc9778b32f0, 0x6f75d1)
	/home/travis/gopath/src/github.com/src-d/borges/vendor/gopkg.in/src-d/go-git.v4/remote.go:319 +0x39 fp=0xc9778b3298 sp=0xc9778b3240 pc=0x8cacd9
github.com/src-d/borges/vendor/gopkg.in/src-d/go-git%2ev4.(*Remote).PushContext(0xc49f4bf7a0, 0xd65620, 0xc61b96d980, 0xc9778b3408, 0x0, 0x0)
	/home/travis/gopath/src/github.com/src-d/borges/vendor/gopkg.in/src-d/go-git.v4/remote.go:76 +0x1b9 fp=0xc9778b33c8 sp=0xc9778b3298 pc=0x8c9209
github.com/src-d/borges.(*temporaryRepository).Push(0xc64e22bc40, 0xd65620, 0xc61b96d980, 0xc52f629620, 0x16, 0xc43233e400, 0x13, 0x20, 0x0, 0x0)
	/home/travis/gopath/src/github.com/src-d/borges/git.go:321 +0x1e3 fp=0xc9778b3460 sp=0xc9778b33c8 pc=0xaa9df3
github.com/src-d/borges.(*Archiver).pushChangesToRootedRepository.func1(0xc52f629620, 0x16, 0xc5efda6ff0, 0xf)
	/home/travis/gopath/src/github.com/src-d/borges/archiver.go:299 +0x1e3 fp=0xc9778b35f8 sp=0xc9778b3460 pc=0xaad393
github.com/src-d/borges.WithInProcRepository(0xc45d0aefc0, 0xc9778b3778, 0x0, 0x0)
	/home/travis/gopath/src/github.com/src-d/borges/git.go:342 +0x3bd fp=0xc9778b36d8 sp=0xc9778b35f8 pc=0xaaa31d
github.com/src-d/borges.(*Archiver).pushChangesToRootedRepository(0xc9778b3e10, 0xd65620, 0xc61b96d980, 0xd692a0, 0xc49f4bec80, 0xc56bb54f20, 0xd60ce0, 0xc64e22bc40, 0x669ee50945ad47d6, 0x9f9e67f81c4d9745, ...)
	/home/travis/gopath/src/github.com/src-d/borges/archiver.go:291 +0x461 fp=0xc9778b3820 sp=0xc9778b36d8 pc=0xaa4001
github.com/src-d/borges.(*Archiver).pushChangesToRootedRepositories(0xc9778b3e10, 0xd65620, 0xc61b96d980, 0xd692a0, 0xc469ed1de0, 0xc82b0a3900, 0xc56bb54f20, 0xd60ce0, 0xc64e22bc40, 0xc45d0ae8d0, ...)
	/home/travis/gopath/src/github.com/src-d/borges/archiver.go:235 +0x70d fp=0xc9778b3ae0 sp=0xc9778b3820 pc=0xaa2bfd
github.com/src-d/borges.(*Archiver).do(0xc9778b3e10, 0xd692a0, 0xc469ed1de0, 0xc82b0a3900, 0x0, 0x0)
	/home/travis/gopath/src/github.com/src-d/borges/archiver.go:168 +0x1229 fp=0xc9778b3d28 sp=0xc9778b3ae0 pc=0xaa1b79
github.com/src-d/borges.(*Archiver).Do(0xc9778b3e10, 0xc82b0a3900, 0xd5e060, 0xca169bae40)
	/home/travis/gopath/src/github.com/src-d/borges/archiver.go:75 +0x14a fp=0xc9778b3da0 sp=0xc9778b3d28 pc=0xaa078a
github.com/src-d/borges.NewArchiverWorkerPool.func1(0xd692a0, 0xc4201f0680, 0xc82b0a3900, 0x0, 0x0)
	/home/travis/gopath/src/github.com/src-d/borges/archiver.go:461 +0x19e fp=0xc9778b3e70 sp=0xc9778b3da0 pc=0xaadd2e
github.com/src-d/borges.(*Worker).Start(0xc4201fe660)
	/home/travis/gopath/src/github.com/src-d/borges/worker.go:45 +0x1d1 fp=0xc9778b3fa8 sp=0xc9778b3e70 pc=0xaac531
github.com/src-d/borges.(*WorkerPool).add.func1(0xc4201f80f0, 0xc4201fe660)
	/home/travis/gopath/src/github.com/src-d/borges/worker_pool.go:77 +0x55 fp=0xc9778b3fd0 sp=0xc9778b3fa8 pc=0xaae415
runtime.goexit()
```

More detailed logs 
[borges-bug-panic-30k.txt](https://github.com/src-d/borges/files/1431649/borges-bug-panic-30k.txt)

**Comment by erizocosmico** [comment]
This should fix the issue: https://github.com/src-d/go-git/pull/639

**Comment by erizocosmico** [comment]
Fixed in v0.8.1.
